### PR TITLE
8293282: LoadLibraryUnloadTest.java fails with "Too few cleared WeakReferences"

### PR DESCRIPTION
**Pull request by RogerRiggs**
Modify the LoadLibraryUnload test to call gc() more a few times, allowing multiple gc cycles to queue the expected refs.
Short the timeout on each cycle to 10 sec.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293282](https://bugs.openjdk.org/browse/JDK-8293282): LoadLibraryUnloadTest.java fails with "Too few cleared WeakReferences"

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10223/head:pull/10223` \
`$ git checkout pull/10223`

Update a local copy of the PR: \
`$ git checkout pull/10223` \
`$ git pull https://git.openjdk.org/jdk pull/10223/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10223`

View PR using the GUI difftool: \
`$ git pr show -t 10223`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10223.diff">https://git.openjdk.org/jdk/pull/10223.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java [hunk 1]
```diff
@@ -162,14 +162,14 @@ public static void main(String[] args) throws Exception {
...
             System.gc();
-            var res = refQueue.remove(Utils.adjustTimeout(30 * 1000L));
-            System.out.println(i + " dequeued: " + res);
-            if (res == null) {
-                Asserts.fail("Too few cleared WeakReferences");
-            }
+            if (refQueue.remove(Utils.adjustTimeout(10 * 1000L)) != null)
+                dequeueCount++;
         }
+        Asserts.assertEquals(dequeueCount, LOADER_COUNT, "Too few cleared WeakReferences");
```
Comment (mlchung) [inline test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:171]: This only invokes GC `LOADER_COUNT` times.  I think you may want something like what ForceGC does:

```
        int retries = (int)(Math.round(1000L * TIMEOUT_FACTOR) / 200);
        for (; retries >= 0; retries--) {
            if (dequeueCount == LOADER_COUNT) return;
            
            System.gc();
            refQueue.remove(200L);
        }
```

**Comment by jaikiran** [comment]
Hello Roger, just a few additional points that I noticed about this test. I haven't fully investigated to see if they play any role at all, but I thought I will note them here:

- The `LoadLibraryUnloadTest` which acts as a driver, launches the `LoadLibraryUnload` as a Java program. There's no specific memory settings passed when that program is launched (line 82 of `LoadLibraryUnloadTest`). I did a quick test a few days back where I updated the process builder launch command to use `-XshowSettings:all` to see what the inferred memory setting would be for this program on our CI systems and I could see that in many cases the estimated max heap size was as big as 7G:
```
[0.001s][trace][gc] MarkStackSize: 4096k  MarkStackSizeMax: 524288k
[0.002s][info ][gc] Using G1
[0.002s][debug][gc] ConcGCThreads: 2 offset 16
[0.002s][debug][gc] ParallelGCThreads: 8
[0.002s][debug][gc] Initialize mark stack with 4096 chunks, maximum 524288
VM settings:
    Max. Heap Size (Estimated): 7.75G
    Using VM: Java HotSpot(TM) 64-Bit Server VM
```
So, would it be relevant/useful if the driver launched the `LoadLibraryUnload` with a more specific/lesser value of max heap size?

- The driver `LoadLibraryUnloadTest` is launched using `-Xcheck:jni` in its test definition. Looking at the code of `LoadLibraryUnloadTest`, I'm unsure why that would be needed on the test definition. I am wondering whether that `-Xcheck:jni` should have instead been passed to the process builder launched `LoadLibraryUnload` process, which is where the actual JNI usage is taking place. Of course this isn't what is causing the test failures.

File: test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java (hunk 1 above)
Comment (RogerRiggs) [reply test/jdk/java/lang/ClassLoader/loadLibraryUnload/LoadLibraryUnload.java:171]: I'll switch to using ForceGC.wait to wait for all the WeakReferences to be cleared.
Using existing logic for the gc and timing.

**Comment by RogerRiggs** [comment]
Hi Jai, I don't think the memory size comes into play, the GC is invoked directly and WeakReferences should be cleared independently from memory pressure.  Adding a memory max might be harmless, but it would not be clear whether it was really a factor.

As for the -Xcheck:jni, it's a left over from when the test was constructed and run as a single main.
There is no explicit JNI usage in the test and the test logic does not depend on that setting.
It should be removed to avoid future confusion.  Thanks for the observations.

**Comment by mlchung** [comment]
Jai has a good observation and setting -Xmx to a smaller heap size.   Less work for GC may shorten the test execution time as it was suspected that this takes more than one GCs to clear the weak reference.

**Review by mlchung — APPROVED** [review]
Looks good.

**Review by jaikiran — APPROVED** [review]
Thank you for the explanation and the changes, Roger.

**Pull request closed without merging**